picard/ui/taglistview.py

```diff
@@ -0,0 +1,223 @@
+        if first_selected_row > 0:
+            self._move_rows_relative(rows, -1)
+
+    def move_selected_rows_down(self):
+        rows = self.get_selected_rows()
+        if not rows:
+            return
+        last_selected_row = max(rows)
+        if last_selected_row < self.model().rowCount() - 1:
+            self._move_rows_relative(rows, 1)
+
+    def select_row(self, row):
+        index = self.model().index(row, 0)
+        self.setCurrentIndex(index)
+
+    def get_selected_rows(self):
+        return [index.row() for index in self.selectedIndexes()]
+
+    def _move_rows_relative(self, rows, direction):
+        model = self.model()
+        current_index = self.currentIndex()
+        selection = self.selectionModel()
+        for row in sorted(rows, reverse=direction > 0):
+            new_index = model.index(row + direction, 0)
+            model.move_row(row, new_index.row())
+            selection.select(new_index, QtCore.QItemSelectionModel.Select)
+            if row == current_index.row():
+                selection.setCurrentIndex(new_index, QtCore.QItemSelectionModel.Current)
+
+
+class TagListModel(QtCore.QAbstractListModel):
+    def __init__(self, tags=[], parent=None):
+        super().__init__(parent)
+        self._tags = [(tag, display_tag_name(tag)) for tag in tags]
+
+    def rowCount(self, parent=QtCore.QModelIndex()):
+        return len(self._tags)
+
+    def data(self, index, role=QtCore.Qt.DisplayRole):
+        if not index.isValid() or role not in (QtCore.Qt.DisplayRole, QtCore.Qt.EditRole):
+            return None
+        i = index.row()
+        field = 1 if role == QtCore.Qt.DisplayRole else 0
+        try:
+            return self._tags[i][field]
+        except IndexError:
+            return None
+
+    def setData(self, index, value, role=QtCore.Qt.EditRole):
+        if not index.isValid() or role not in (QtCore.Qt.DisplayRole, QtCore.Qt.EditRole):
+            return False
+        i = index.row()
+        try:
+            if role == QtCore.Qt.EditRole:
+                display_name = display_tag_name(value) if value else value
+                self._tags[i] = (value, display_name)
+            elif role == QtCore.Qt.DisplayRole:
+                current = self._tags[i]
+                self._tags[i] = (current[0], value)
+            return True
+        except IndexError:
+            return False
+
+    def flags(self, index):
+        if index.isValid():
+            return (QtCore.Qt.ItemIsSelectable
+                | QtCore.Qt.ItemIsEditable
+                | QtCore.Qt.ItemIsDragEnabled
+                | QtCore.Qt.ItemIsEnabled
+                | QtCore.Qt.ItemNeverHasChildren)
+        else:
+            return QtCore.Qt.ItemIsDropEnabled
+
+    def insertRows(self, row, count, parent=QtCore.QModelIndex()):
+        super().beginInsertRows(parent, row, row + count - 1)
+        for i in range(count):
+            self._tags.insert(row, ("", ""))
```

Comment:
I'm bothered by this loop.
What about, something like:

```python
self._tags = self._tags[:row] + [("", "")] * count + self._tags[row:]
```
I didn't test, nor did any speed comparaison, but it gets rid of unused loop variable, and match the style of what you did in `removeRows()` just below.

Reply:
Not a big fan of this to be honest, I find this list construction not very readable. And if this is about performance: In practice this loop won't run more than once since we don't have any way to insert multiple tags. It's just the interface that requires this implementation.